### PR TITLE
Scaffold controller

### PR DESCRIPTION
**Pull request by jakkab**
Changes:
- commit 1: 
  - scaffold controller
- commit 2:
  - api
  - basic controller logic
  - hydra client CRUD
  - validation
  - happy path integration test
  - enable ci

### CODE REVIEW

File: api/v1alpha1/oauth2client_types.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OAuth2ClientSpec defines the desired state of OAuth2Client
+type OAuth2ClientSpec struct {
+	// GrantTypes is an array of grant types the client is allowed to use.
+	//
+	// Pattern: client_credentials|authorization_code|implicit|refresh_token
```
Comment (Demonsthere) [inline api/v1alpha1/oauth2client_types.go]: If we already have this so well defined, maybe use the kubebuilder [verification annotations](https://book.kubebuilder.io/reference/markers/crd-validation.html) to include this in the CRD?

File: config/crd/kustomization.yaml [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# This kustomization.yaml is not intended to be run by itself,
+# since it depends on service name and namespace that are out of this kustomize package.
+# It should be run by config/default
+resources:
+- bases/hydra.ory.sh_oauth2clients.yaml
+# +kubebuilder:scaffold:crdkustomizeresource
+
+patches:
```
Comment (Demonsthere) [inline config/crd/kustomization.yaml]: In the current version of `patches` is `patchesStrategicMerge`, please update field name in all `kustomization.yaml` files

File: config/default/manager_image_patch.yaml-e [hunk 3]
```diff
@@ -0,0 +1,12 @@
+apiVersion: apps/v1
```
Comment (Demonsthere) [inline config/default/manager_image_patch.yaml-e]: This file seems like a sed artifact

File: api/v1alpha1/oauth2client_types.go (hunk 1 above)
Comment (jakkab) [reply api/v1alpha1/oauth2client_types.go]: validation added!

File: config/crd/kustomization.yaml (hunk 2 above)
Comment (jakkab) [reply config/crd/kustomization.yaml]: regenerated using kubebuilder v2.0.0.

File: config/default/manager_image_patch.yaml-e (hunk 3 above)
Comment (jakkab) [reply config/default/manager_image_patch.yaml-e]: removed.

File: api/v1alpha1/oauth2client_json.go [hunk 4]
```diff
@@ -0,0 +1,11 @@
+package v1alpha1
+
+// OAuth2ClientJSON represents an OAuth2 client digestible by ORY Hydra
+type OAuth2ClientJSON struct {
```
Comment (Tomasz-Smelcerz-SAP) [inline api/v1alpha1/oauth2client_json.go]: This type does not really belong here. It's not tied to the v1alpha1 version at all. I would move it to top-level "hydra" package

File: controllers/oauth2client_controller.go [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+	return ctrl.Result{}, nil
+}
+
+func (r *OAuth2ClientReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&hydrav1alpha1.OAuth2Client{}).
+		Complete(r)
+}
+
+func (r *OAuth2ClientReconciler) registerOAuth2Client(ctx context.Context, client *hydrav1alpha1.OAuth2Client) error {
```
Comment (Tomasz-Smelcerz-SAP) [inline controllers/oauth2client_controller.go]: Consider extracting this to a "Client" struct in a top-level "hydra" package

File: controllers/oauth2client_controller_test.go [hunk 6]
```diff
@@ -0,0 +1,227 @@
...
+	emptyBody         = `{}`
+)
+
+var testOAuthJSON = &v1alpha1.OAuth2ClientJSON{
+	Name:       "test-name-2",
+	Scope:      "some,other,scopes",
+	GrantTypes: []string{"type2"},
+}
+
+func TestCRUD(t *testing.T) {
```
Comment (Tomasz-Smelcerz-SAP) [inline controllers/oauth2client_controller_test.go]: Contrary to the name, this test does not test the controller but the rest client. It should be moved along with the client to the "hydra" package

File: controllers/oauth2client_controller.go [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+	Log        logr.Logger
+	HTTPClient *http.Client
+	client.Client
+}
+
+// +kubebuilder:rbac:groups=hydra.ory.sh,resources=oauth2clients,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=hydra.ory.sh,resources=oauth2clients/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups="",resources=secrets,verbs=get;list;watch;create;update;patch;delete
+
+func (r *OAuth2ClientReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
```
Comment (Tomasz-Smelcerz-SAP) [inline controllers/oauth2client_controller.go]: We should have a basic integration test for the "happy path" scenario.
I imagine the following flow:
1) User creates invalid CR instance
- User sees proper response (either from the Controller or the API server).
- There is no call to Hydra
2) User creates valid CR instance
- The controller makes a call to Hydra (test mock)
- User sees valid status on CR

File: api/v1alpha1/oauth2client_json.go (hunk 4 above)
Comment (jakkab) [reply api/v1alpha1/oauth2client_json.go]: Sure thing, will be moved.

File: controllers/oauth2client_controller.go (hunk 5 above)
Comment (jakkab) [reply controllers/oauth2client_controller.go]: good point!

File: controllers/oauth2client_controller_test.go (hunk 6 above)
Comment (jakkab) [reply controllers/oauth2client_controller_test.go]: yes sir!

File: controllers/oauth2client_controller.go (hunk 7 above)
Comment (jakkab) [reply controllers/oauth2client_controller.go]: Will be addressed in a separate PR.

**Pull request merged**